# 01 - Data Loading
**Project:** Global Food Security Risk Analysis  
**Phase:** 1 - Data Collection & Understanding  
**Goal:** Load all raw datasets, inspect structure, and confirm readiness for cleaning.

---

## Data Sources
| File | Source | Description |
|---|---|---|
| `FAOSTAT_FoodSecurity.csv` | FAOSTAT | Suite of Food Security Indicators |
| `FAOSTAT_FoodBalances.csv` | FAOSTAT | Food Balances (kcal, protein, fat per capita) |
| `FAOSTAT_ProductionIndices.csv` | FAOSTAT | Gross per capita Production Index |
| `WorldBank_WDI_Data.csv` | World Bank | GDP, population, precipitation, poverty |

## 0. Setup

In [10]:
import pandas as pd
import numpy as np
import os

# Display settings
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', 50)
pd.set_option('display.float_format', '{:.2f}'.format)

# Data paths
RAW_DATA_PATH = '../data/raw/'

print('Setup complete.')
print(f'Files in raw data folder:')
for f in os.listdir(RAW_DATA_PATH):
    print(f'  {f}')

Setup complete.
Files in raw data folder:
  FoodBalances.csv
  ProductIonIndices.csv
  Suite_of_Food_Security_Indicators.csv
  WorldDevelopmentIndicators.csv


---
## 1. FAOSTAT — Suite of Food Security Indicators

In [11]:
fao_fs = pd.read_csv(RAW_DATA_PATH + 'Suite_of_Food_Security_Indicators.csv')

print('=== Shape ===')
print(fao_fs.shape)

print('\n=== Columns ===')
print(fao_fs.columns.tolist())

print('\n=== Head ===')
fao_fs.head()

=== Shape ===
(20471, 15)

=== Columns ===
['Domain Code', 'Domain', 'Area Code (M49)', 'Area', 'Element Code', 'Element', 'Item Code', 'Item', 'Year Code', 'Year', 'Unit', 'Value', 'Flag', 'Flag Description', 'Note']

=== Head ===


,Domain Code,Domain,Area Code (M49),Area,Element Code,Element,Item Code,Item,Year Code,Year,Unit,Value,Flag,Flag Description,Note
0,FS,Suite of Food Security Indicators,4,Afghanistan,6128,Value,220001,Dietary energy supply used in the estimation o...,2010,2010,kcal/cap/d,2200,E,Estimated value,NaN
1,FS,Suite of Food Security Indicators,4,Afghanistan,6128,Value,220001,Dietary energy supply used in the estimation o...,2011,2011,kcal/cap/d,2172,E,Estimated value,NaN
2,FS,Suite of Food Security Indicators,4,Afghanistan,6128,Value,220001,Dietary energy supply used in the estimation o...,2012,2012,kcal/cap/d,2166,E,Estimated value,NaN
3,FS,Suite of Food Security Indicators,4,Afghanistan,6128,Value,220001,Dietary energy supply used in the estimation o...,2013,2013,kcal/cap/d,2205,E,Estimated value,NaN
4,FS,Suite of Food Security Indicators,4,Afghanistan,6128,Value,220001,Dietary energy supply used in the estimation o...,2014,2014,kcal/cap/d,2270,E,Estimated value,NaN


In [12]:
# Unique items (indicators)
print('=== Unique Items (Indicators) ===')
print(fao_fs['Item'].unique())

print('\n=== Year Range ===')
print(f"Min: {fao_fs['Year'].min()}, Max: {fao_fs['Year'].max()}")

print('\n=== Number of Countries ===')
print(fao_fs['Area'].nunique())

=== Unique Items (Indicators) ===
['Dietary energy supply used in the estimation of the prevalence of undernourishment (kcal/cap/day)'
 'Gross domestic product per capita, PPP, (constant 2021 international $)'
 'Prevalence of undernourishment (percent) (3-year average)'
 'Prevalence of moderate or severe food insecurity in the total population (percent) (3-year average)'
 'Prevalence of moderate or severe food insecurity in the male adult population (percent) (3-year average)'
 'Prevalence of moderate or severe food insecurity in the female adult population (percent) (3-year average)'
 'Cereal import dependency ratio (percent) (3-year average)'
 'Political stability and absence of violence/terrorism (index)'
 'Per capita food supply variability (kcal/cap/day)']

=== Year Range ===
Min: 2009-2011, Max: 2023

=== Number of Countries ===
204


In [13]:
# Missing values
print('=== Missing Values ===')
print(fao_fs.isnull().sum())

print('\n=== Missing Rate for Value column ===')
missing_rate = fao_fs['Value'].isnull().sum() / len(fao_fs) * 100
print(f'{missing_rate:.1f}%')

=== Missing Values ===
Domain Code             0
Domain                  0
Area Code (M49)         0
Area                    0
Element Code            0
Element                 0
Item Code               0
Item                    0
Year Code               0
Year                    0
Unit                 2535
Value                2588
Flag                    0
Flag Description        0
Note                16954
dtype: int64

=== Missing Rate for Value column ===
12.6%


---
## 2. FAOSTAT — Food Balances

In [14]:
fao_fb = pd.read_csv(RAW_DATA_PATH + 'FoodBalances.csv')

print('=== Shape ===')
print(fao_fb.shape)

print('\n=== Columns ===')
print(fao_fb.columns.tolist())

print('\n=== Head ===')
fao_fb.head()

=== Shape ===
(7275, 15)

=== Columns ===
['Domain Code', 'Domain', 'Area Code (M49)', 'Area', 'Element Code', 'Element', 'Item Code (FBS)', 'Item', 'Year Code', 'Year', 'Unit', 'Value', 'Flag', 'Flag Description', 'Note']

=== Head ===


,Domain Code,Domain,Area Code (M49),Area,Element Code,Element,Item Code (FBS),Item,Year Code,Year,Unit,Value,Flag,Flag Description,Note
0,FBS,Food Balances (2010-),4,Afghanistan,664,Food supply (kcal/capita/day),S2901,Grand Total,2010,2010,kcal/cap/d,2200.21,E,Estimated value,NaN
1,FBS,Food Balances (2010-),4,Afghanistan,674,Protein supply quantity (g/capita/day),S2901,Grand Total,2010,2010,g/cap/d,65.54,E,Estimated value,NaN
2,FBS,Food Balances (2010-),4,Afghanistan,684,Fat supply quantity (g/capita/day),S2901,Grand Total,2010,2010,g/cap/d,37.93,E,Estimated value,NaN
3,FBS,Food Balances (2010-),4,Afghanistan,664,Food supply (kcal/capita/day),S2901,Grand Total,2011,2011,kcal/cap/d,2171.86,E,Estimated value,NaN
4,FBS,Food Balances (2010-),4,Afghanistan,674,Protein supply quantity (g/capita/day),S2901,Grand Total,2011,2011,g/cap/d,63.74,E,Estimated value,NaN


In [15]:
# Unique elements
print('=== Unique Elements ===')
print(fao_fb['Element'].unique())

print('\n=== Missing Rate for Value column ===')
missing_rate = fao_fb['Value'].isnull().sum() / len(fao_fb) * 100
print(f'{missing_rate:.1f}%')

=== Unique Elements ===
['Food supply (kcal/capita/day)' 'Protein supply quantity (g/capita/day)'
 'Fat supply quantity (g/capita/day)']

=== Missing Rate for Value column ===
0.0%


---
## 3. FAOSTAT — Production Indices

In [16]:
fao_qi = pd.read_csv(RAW_DATA_PATH + 'ProductionIndices.csv')

print('=== Shape ===')
print(fao_qi.shape)

print('\n=== Columns ===')
print(fao_qi.columns.tolist())

print('\n=== Head ===')
fao_qi.head()

=== Shape ===
(4506, 14)

=== Columns ===
['Domain Code', 'Domain', 'Area Code (M49)', 'Area', 'Element Code', 'Element', 'Item Code (CPC)', 'Item', 'Year Code', 'Year', 'Unit', 'Value', 'Flag', 'Flag Description']

=== Head ===


,Domain Code,Domain,Area Code (M49),Area,Element Code,Element,Item Code (CPC),Item,Year Code,Year,Unit,Value,Flag,Flag Description
0,QI,Production Indices,4,Afghanistan,434,Gross per capita Production Index Number (2014...,F2051,Agriculture,2010,2010,NaN,109.33,E,Estimated value
1,QI,Production Indices,4,Afghanistan,434,Gross per capita Production Index Number (2014...,F2051,Agriculture,2011,2011,NaN,100.63,E,Estimated value
2,QI,Production Indices,4,Afghanistan,434,Gross per capita Production Index Number (2014...,F2051,Agriculture,2012,2012,NaN,107.26,E,Estimated value
3,QI,Production Indices,4,Afghanistan,434,Gross per capita Production Index Number (2014...,F2051,Agriculture,2013,2013,NaN,101.02,E,Estimated value
4,QI,Production Indices,4,Afghanistan,434,Gross per capita Production Index Number (2014...,F2051,Agriculture,2014,2014,NaN,104.83,E,Estimated value


In [17]:
# Unique items
print('=== Unique Items ===')
print(fao_qi['Item'].unique())

print('\n=== Year Range ===')
print(f"Min: {fao_qi['Year'].min()}, Max: {fao_qi['Year'].max()}")

print('\n=== Missing Rate for Value column ===')
missing_rate = fao_qi['Value'].isnull().sum() / len(fao_qi) * 100
print(f'{missing_rate:.1f}%')

=== Unique Items ===
['Agriculture' 'Food']

=== Year Range ===
Min: 2010, Max: 2023

=== Missing Rate for Value column ===
0.0%


---
## 4. World Bank — World Development Indicators

In [18]:
# World Bank uses '..' for missing values
wb = pd.read_csv(
    RAW_DATA_PATH + 'WorldDevelopmentIndicators.csv',
    na_values=['..']  # convert '..' to NaN
)

print('=== Shape ===')
print(wb.shape)

print('\n=== Columns ===')
print(wb.columns.tolist())

print('\n=== Head ===')
wb.head()

=== Shape ===
(1601, 17)

=== Columns ===
['Country Name', 'Country Code', 'Series Name', 'Series Code', '2010 [YR2010]', '2011 [YR2011]', '2012 [YR2012]', '2013 [YR2013]', '2014 [YR2014]', '2015 [YR2015]', '2016 [YR2016]', '2017 [YR2017]', '2018 [YR2018]', '2019 [YR2019]', '2020 [YR2020]', '2021 [YR2021]', '2022 [YR2022]']

=== Head ===


,Country Name,Country Code,Series Name,Series Code,2010 [YR2010],2011 [YR2011],2012 [YR2012],2013 [YR2013],2014 [YR2014],2015 [YR2015],2016 [YR2016],2017 [YR2017],2018 [YR2018],2019 [YR2019],2020 [YR2020],2021 [YR2021],2022 [YR2022]
0,Afghanistan,AFG,GDP per capita (current US$),NY.GDP.PCAP.CD,560.62,606.69,651.42,637.09,625.05,565.57,522.08,525.47,491.34,496.60,510.79,356.50,357.26
1,Afghanistan,AFG,Poverty headcount ratio at $3.00 a day (2021 P...,SI.POV.DDAY,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,Afghanistan,AFG,"Population, total",SP.POP.TOTL,28284089.00,29347708.00,30560034.00,31622704.00,32792523.00,33831764.00,34700612.00,35688935.00,36743039.00,37856121.00,39068979.00,40000412.00,40578842.00
3,Afghanistan,AFG,Population growth (annual %),SP.POP.GROW,2.93,3.69,4.05,3.42,3.63,3.12,2.54,2.81,2.91,2.98,3.15,2.36,1.44
4,Afghanistan,AFG,Average precipitation in depth (mm per year),AG.LND.PRCP.MM,327.00,327.00,327.00,327.00,327.00,327.00,327.00,327.00,327.00,327.00,327.00,327.00,327.00


In [19]:
# Unique series (indicators)
print('=== Unique Series (Indicators) ===')
print(wb['Series Name'].unique())

print('\n=== Number of Countries ===')
print(wb['Country Name'].nunique())

=== Unique Series (Indicators) ===
['GDP per capita (current US$)'
 'Poverty headcount ratio at $3.00 a day (2021 PPP) (% of population)'
 'Population, total' 'Population growth (annual %)'
 'Average precipitation in depth (mm per year)'
 'Droughts, floods, extreme temperatures (% of population, average 1990-2009)'
 nan]

=== Number of Countries ===
268


In [21]:
# Missing rate per indicator
print('=== Missing Rate per Indicator ===')
year_cols = [c for c in wb.columns if 'YR' in c]
for series in wb['Series Name'].dropna().unique():
    subset = wb[wb['Series Name'] == series][year_cols]
    missing = subset.isnull().sum().sum()
    total = subset.size
    print(f'{str(series)[:50]}: {missing/total*100:.1f}% missing')

=== Missing Rate per Indicator ===
GDP per capita (current US$): 2.5% missing
Poverty headcount ratio at $3.00 a day (2021 PPP) : 64.5% missing
Population, total: 0.4% missing
Population growth (annual %): 0.4% missing
Average precipitation in depth (mm per year): 31.8% missing
Droughts, floods, extreme temperatures (% of popul: 100.0% missing


---
## 5. Summary

| Dataset | Shape | Year Range | Countries | Notes |
|---|---|---|---|---|
| Food Security Indicators | (20471, 15) | 2009-2011 ~ 2023 | 204 | 3-year average format, 12.6% missing |
| Food Balances | (7275, 15) | 2010-2022 | 179| Remove kcal (duplicate), 0% missing |
| Production Indices | (4506, 14) | 2010-2023 | 200 | Drop 2023 rows, 0% missing |
| World Bank WDI | (1601, 17) | 2010-2022 | 268 | Drop Poverty & Droughts (high missing) |

**Next Step:** `02_cleaning.ipynb`
- Standardize country names / codes
- Handle missing values
- Filter years to 2010-2022
- Pivot to wide format

In [22]:
print(f"Food Balances countries: {fao_fb['Area'].nunique()}")
print(f"Production Indices countries: {fao_qi['Area'].nunique()}")

Food Balances countries: 179
Production Indices countries: 200
